Refactor arguments checks for foraech APIs

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46664 Move device guard from MultiTensorApply.cuh
* #46663 Remove .pyi from /_multi_tensor
* #46662 Renamed a TensorListMetaData property. Cleaned up a test
* **#46661 Refactor arguments checks for foraech APIs**

Differential Revision: [D24453342](https://our.internmc.facebook.com/intern/diff/D24453342)


**In this PR**
Refactored argument checks for foreach APIs

Tested via unit tests

Commit:
Optimize arguments checks

[ghstack-poisoned]

Commit:
Update on "Optimize arguments checks"




[ghstack-poisoned]

Commit:
Update on "Optimize arguments checks"


Differential Revision: [D24453342](https://our.internmc.facebook.com/intern/diff/D24453342)

[ghstack-poisoned]

Commit:
Update on "Optimize arguments checks"


Differential Revision: [D24453342](https://our.internmc.facebook.com/intern/diff/D24453342)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/ForeachBinaryOpList.cu:
```diff
@@ -51,8 +51,6 @@ void foreach_tensor_list_op_(TensorList tensors1, TensorList tensors2, Scalar al
 
 #define FOREACH_BINARY_OP_LIST(NAME, OP)                                                                    \
 void foreach_tensor_##NAME##_list_kernel_cuda_(TensorList tensors1, TensorList tensors2) {                  \
-    check_foreach_api_restrictions(tensors1, tensors2);                                                     \
```
is check_foreach_api_restrictions called anywhere now?

izdeby replied:
Yeah, in slow path. [ForeachOpsKernels.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ForeachOpsKernels.cpp#L8)

izdeby committed:
Update on "Optimize arguments checks"


Differential Revision: [D24453342](https://our.internmc.facebook.com/intern/diff/D24453342)


**In this PR**
Optimize the API checks. Instead of going multiple times over the same inputs and checking for different things like size, dtype, device, etc - do all that but in one run. CPU perf improvement. 

Tested via unit tests

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -188,12 +201,66 @@ bool can_use_fast_route(TensorList tensors1, TensorList tensors2, TensorList ten
       return false;
     }
 
+    if (tensors1[i].strides() != tensors2[i].strides() ||
```
There's a lot of repetition between can_use_fast_route(TL, TL), can_use_fast_route(TL, TL, TL), and can_use_fast_route(TL, TL, TL, Scalar) to check that "each respective Tensor has the same attributes" (for some definition of same attributes).

Would a helper function `has_same_attributes(TensorList tensors)` help? Then we could implement
```
bool can_use_fast_route(TensorList first, TensorList other) {
   for (int64_t i = 0; ...) {
       if not has_same_attributes({first[i], other[i]}) {
           return false;
       }
   }
   return true;
}

bool can_use_fast_route(TensorList first, TensorList second, TensorList third,) {
   for (int64_t i = 0; ...) {
       if not has_same_attributes({first[i], second[i], third[i]}) {
           return false;
       }
   }
   return true;
}

bool can_use_fast_route(TensorList first, TensorList second, TensorList third, Scalar fourth) {
   for (int64_t i = 0; ...) {
       if not has_same_attributes({first[i], second[i], third[i]}) {
           return false;
       }
       // check type promotion with Scalar fourth
       ...
   }
   return true;
}
```

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -164,17 +168,26 @@ bool can_use_fast_route(TensorList tensors1, TensorList tensors2) {
 
-bool can_use_fast_route(TensorList tensors1, TensorList tensors2, TensorList tensors3) {
+bool can_use_fast_route(TensorList tensors1, TensorList tensors2, TensorList tensors3, Scalar scalar) {
+  TORCH_CHECK(tensors1.size() > 0, "Tensor list must have at least one tensor.");
+  TORCH_CHECK(tensors1.size() == tensors2.size(), "Tensor lists must have the same number of tensors, got ", tensors1.size(), " and ", tensors2.size());
+  TORCH_CHECK(tensors2.size() == tensors3.size(), "Tensor lists must have the same number of tensors, got ", tensors2.size(), " and ", tensors3.size());
+
 #ifdef __HIP_PLATFORM_HCC__
   return false;
 #else
   auto expected_device = tensors1[0].device();
+  auto expected_dtype = tensors1[0].dtype();
+
   for (int64_t i = 0; i < tensors1.size(); i++) {
     TORCH_CHECK(tensors1[i].sizes() == tensors2[i].sizes(), 
                 "Corresponding tensors from tensor lists have different sizes, got ", tensors1[i].sizes(), " and ", tensors2[i].sizes());
 
     TORCH_CHECK(tensors1[i].sizes() == tensors3[i].sizes(), 
                 "Corresponding tensors from tensor lists have different sizes, got ", tensors1[i].sizes(), " and ", tensors3[i].sizes());
+    TORCH_CHECK(tensors1[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors2[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors3[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
```
It feels a little weird to me that we're doing some error checking in a function named "can_use_fast_route": I'd expect "can_use_fast_route" to just return a True/False and never error out.

ngimel replied:
That's because can_use_fast_route is now doing double duty of `check_foreach_api_restrictions` and actually checking fast route. I agree it's a bit confusing, @izdeby what's the motivation behind this?

izdeby replied:
I know. but i dont wanna iterate over the lists twice where first time will be checking for the hard requirements and error out and 2nd time over slow/fast path.

izdeby replied:
Im trying to reduce CPU time. This is the main goal.

zou3519 replied:
How different are the numbers? I wouldn't expect checking the dtypes for these things to add very much to the overall runtime

izdeby replied:
@zou3519, @ngimel returned `check_foreach_api_restrictions` and just did some refactoring and cleanup.

Commit:
Update on "Optimize arguments checks"


Differential Revision: [D24453342](https://our.internmc.facebook.com/intern/diff/D24453342)


**In this PR**
Optimize the API checks. Instead of going multiple times over the same inputs and checking for different things like size, dtype, device, etc - do all that but in one run. CPU perf improvement. 

Tested via unit tests

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -73,51 +72,63 @@ bool can_use_fast_route(TensorList tensors, Scalar scalar) {
-    if (scalar.isIntegral(/*includeBool*/ false) && t.dtype() == at::kBool) {
-      return false;
-    }
+bool tensor_and_scalar_are_same_type(Tensor tensor, Scalar scalar) {
```
This isn't really checking if they're the same type, right? A float32 Tensor and a float64 Scalar would pass this check. It's more like it is checking that we don't end up promoting `tensor`. Maybe something like `bool will_promote_tensor(const Tensor& tensor, Scalar scalar)`

zou3519 replied:
(also remember to take Tensor by const ref, intrusive_ptr refcount bumps are relatively expensive!)

zou3519 commented on aten/src/ATen/native/ForeachUtils.h:
```diff
@@ -44,27 +32,38 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2) {
 
-  for (auto t : tensors) {
+// Check if all tensors have the same device, layout, strides and are not overlapping and dense
+bool has_same_attributes(Device expected_device, TensorList tensors) {
```
Since this is in the at::native namespace, we should probably name it something more descriptive to avoid confusion.

zou3519 replied:
EDIT: nvm, I see that this is enclosed in an anonymous namespace.

Commit:
Update on "Refactor arguments checks for foraech APIs"


Differential Revision: [D24453342](https://our.internmc.facebook.com/intern/diff/D24453342)


**In this PR**
Refactored argument checks for foreach APIs

Tested via unit tests

[ghstack-poisoned]
